File: LCABSTRecursive.java
//TC: O(n) n is the number of nodes in the tree
//SC: O(H) - height of tree recursive stack space

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        
        if(root == null ) return null;
        
        if(root.val < p.val && root.val < q.val)
            return lowestCommonAncestor(root.right,p,q);
        else if(root.val > p.val && root.val > q.val)
            return lowestCommonAncestor(root.left,p,q);
        else{
            return root;
        }
        
    }
}

File: LcaBSTRecursive.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# TC - O(n)
# SC - O(n)
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':

        def lcaRecursive(node):
            if node is None or node == p or node == q:
                return node
            left = lcaRecursive(node.left)
            right = lcaRecursive(node.right)

            if not left and not right:
                return None
            elif right and not left:
                return right
            elif left and not right:
                return left
            else:
                return node
        return lcaRecursive(root)
